### PR TITLE
Add linalg directory and initialization code.

### PR DESCRIPTION
**Pull request by alewis**

### CODE REVIEW

File: tensornetwork/linalg/linalg.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    fname:  Name of the method of backend to call (a string).
+    *fargs: Positional arguments to the  initialization method.
+    name: Optional name of the Node.
+    axis_names: Optional names of the Node's dangling edges.
+    backend: The backend or its name.
+    **fkwargs: Keyword arguments to the initialization method.
+
+  Returns:
+    node: A Node wrapping data generated by
+          (the_backend).fname(*fargs, **kwargs), with one dangling edge per
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:54]: typo

File: tensornetwork/linalg/linalg.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                             dtype=dtype, M=M)
+  return the_node
+
+
+def zeros(shape: Sequence[int],
+          dtype: Optional[Type[np.number]],
+          name: Optional[Text] = None,
+          axis_names: Optional[List[Text]] = None,
+          backend: Optional[Union[Text, BaseBackend]] = None) -> Tensor:
+  """Return a Node of shape 'shape' of all zeros.
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:100]: use `shape` instead of 'shape'

File: tensornetwork/linalg/linalg.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        ones.
+  """
+  the_node = initialize_node("eye", N,
+                             name=name, axis_names=axis_names, backend=backend,
+                             dtype=dtype, M=M)
+  return the_node
+
+
+def zeros(shape: Sequence[int],
+          dtype: Optional[Type[np.number]],
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:96]: should this have a default value?

File: tensornetwork/linalg/linalg.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    backend_obj = backends.backend_factory.get_backend(backend)
+  func = getattr(backend_obj, fname)
+  data = func(*fargs, **fkwargs)
+  node = network_components.Node(data, name=name, axis_names=axis_names,
+                                 backend=backend)
+  return node
+
+
+def eye(N: int,
+        dtype: Optional[Type[np.number]],
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:69]: default value?

File: tensornetwork/linalg/linalg.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        Represents an array of all zeros.
+  """
+  the_node = initialize_node("zeros", shape,
+                             name=name, axis_names=axis_names, backend=backend,
+                             dtype=dtype)
+  return the_node
+
+
+def ones(shape: Sequence[int],
+         dtype: Optional[Type[np.number]],
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:119]: default

File: tensornetwork/linalg/linalg.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+         name: Optional[Text] = None,
+         axis_names: Optional[List[Text]] = None,
+         backend: Optional[Union[Text, BaseBackend]] = None) -> Tensor:
+  """Return a Node of shape 'shape' of all ones.
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:123]: quotes

File: tensornetwork/linalg/linalg.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        Represents an array of all ones.
+  """
+  the_node = initialize_node("ones", shape,
+                             name=name, axis_names=axis_names, backend=backend,
+                             dtype=dtype)
+  return the_node
+
+
+def randn(shape: Sequence[int],
+          dtype: Optional[Type[np.number]],
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:142]: default

File: tensornetwork/linalg/linalg.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+          seed: Optional[int] = None,
+          name: Optional[Text] = None,
+          axis_names: Optional[List[Text]] = None,
+          backend: Optional[Union[Text, BaseBackend]] = None) -> Tensor:
+  """Return a Node of shape 'shape' of Gaussian random floats.
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:147]: quotes

File: tensornetwork/linalg/linalg.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        Represents an array of all ones.
+  """
+  the_node = initialize_node("randn", shape,
+                             name=name, axis_names=axis_names, backend=backend,
+                             seed=seed, dtype=dtype)
+  return the_node
+
+
+def random_uniform(shape: Sequence[int],
+                   dtype: Optional[Type[np.number]],
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:167]: dtype

File: tensornetwork/linalg/linalg.py (hunk 3 above)
Comment (alewis) [reply tensornetwork/linalg/linalg.py:96]: That seems reasonable, yes. What should it be?

File: tensornetwork/linalg/linalg.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                   seed: Optional[int] = None,
+                   boundaries: Optional[Tuple[float, float]] = (0.0, 1.0),
```
Comment (mganahl) [inline tensornetwork/linalg/linalg.py:169]: apparently `Optional` means that it can be `None` as well. If that is not possible then  remove `Optional`

File: tensornetwork/linalg/linalg.py (hunk 3 above)
Comment (mganahl) [reply tensornetwork/linalg/linalg.py:96]: I would say double precision floats

File: tensornetwork/linalg/linalg_test.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+  name = "Jeffrey"
+  axis_names = ["Sam", "Blinkey"]
+  backend_list = ["jax", "numpy"]
+  for backend in backend_list:
+    for dtype in np_dtypes + np_int_dtypes + np_uint_dtypes:
+      tnI = linalg.eye(N, dtype=dtype, M=M, name=name, axis_names=axis_names,
+                       backend=backend)
+      npI = np.eye(N, M=M, dtype=dtype)
+      np.testing.assert_allclose(tnI.tensor, npI)
+
```
Comment (mganahl) [inline tensornetwork/linalg/linalg_test.py:33]: also check if axis names are as expected

Comment (mganahl) [reply tensornetwork/linalg/linalg_test.py:33]: and `name`

Comment (mganahl) [reply tensornetwork/linalg/linalg_test.py:33]: and backend

Comment (mganahl) [reply tensornetwork/linalg/linalg_test.py:33]: also in all other tests

Comment (mganahl) [reply tensornetwork/linalg/linalg_test.py:33]: to test all backends you can use a @pytest.mark.parametrize(backend=['numpy','jax','tensorflow','pytorch']), or just use the backend fixture `backend`, i.e.

```python 
test_eye(backend): #the variable `backend` is defined by the pytest fixture
  ...

```

Comment (mganahl) [reply tensornetwork/linalg/linalg_test.py:33]: In general we always test all backends (except `symmetric`)

File: tensornetwork/linalg/linalg.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+from tensornetwork import network_components
+
+
+Tensor = Any
+BaseBackend = base_backend.BaseBackend
+
+
+# INITIALIZATION
+def initialize_node(fname: Text,
+                    *fargs: Any,
```
Comment (Thenerdstation) [inline tensornetwork/linalg/linalg.py:33]: Looks like you never actually use these positional args. Can you remove it in favor of just always using `fkwargs`?

Comment (alewis) [reply tensornetwork/linalg/linalg.py:33]: I use them to feed in the shape in a couple of places, but it's easy enough to make that a keyword argument if you like that more

Comment (Thenerdstation) [reply tensornetwork/linalg/linalg.py:33]: Yeah that sounds good. More readable that way too.

Comment (alewis) [reply tensornetwork/linalg/linalg.py:33]: Ah, I remember why I did this now; it's because in certain cases the backend function expects certain arguments to be positional. If only `fkwargs` is used I don't know how to determine which are which.

Comment (Thenerdstation) [reply tensornetwork/linalg/linalg.py:33]: Aw, ok no worries then. Keep it as-is.

File: tensornetwork/linalg/linalg.py (hunk 2 above)
Comment (alewis) [reply tensornetwork/linalg/linalg.py]: Do you mean the word shape or the Markdown-formatted `shape`

File: tensornetwork/linalg/linalg.py (hunk 3 above)
Comment (alewis) [reply tensornetwork/linalg/linalg.py]: The default is now None, under the (tested) assumption that all backends provide a default value in this case.

Comment (mganahl) [reply tensornetwork/linalg/linalg.py]: why not np.float64?

Comment (mganahl) [reply tensornetwork/linalg/linalg.py]: then it would be clearer for the user

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: Will this be valid for all backends (notably pytorch)?

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: It also strikes me that the "best default" ought to be located at the backend; one could imagine that different backends might prefer different dtypes.

Comment (mganahl) [reply tensornetwork/linalg/linalg.py]: I see, passing numpy dtypes using the backend initialization is problematic, yes. The default value though should be set within the top level function, and the user should be able to know what it is without looking into the backend documentation. Otherwise this is could be a source of really obscure bugs.

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: Well, I guess I could just have a big if statement or something. For that to work, Pytorch would need to be imported at the top level, which would make it a dependency of the interface. Is that something we want?

Comment (mganahl) [reply tensornetwork/linalg/linalg.py]: No we don't want that. My suggestion is to use numpy dtypes and convert them to the backend dtypes within the backend. Maybe I'm too picky, I just think that hidden choices about dtypes are really annoying.

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: I'm fine with that, but then wouldn't the user need to consult the backend to figure out that the conversion had occurred?

Comment (mganahl) [reply tensornetwork/linalg/linalg.py]: The idea is to use numpy to configure the dtypes (i.e. half vs single vs double), and convert this to the corresponding backend dtypes, (ie.e. np.floa64 will be converted to jax.numpy.float64, or torch.float64)

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: We could give each backend a member "backend.default_dtype". Then, it would be obvious that each backend might do something different (rather than users perhaps assuming that their Nodes literally store np.float64), and while one would still need to look up the backend documentation to figure out what the default was, at least there would be only one clear place to look.

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: @Thenerdstation

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: > The idea is to use numpy to configure the dtypes (i.e. half vs single vs double), and convert this to the corresponding backend dtypes, (ie.e. np.floa64 will be converted to jax.numpy.float64, or torch.float64)

This also seems like a reasonable solution, provided it be documented appropriately.

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: Jax, NumPy, and TensorFlow already behave in this way, so it seems only Pytorch would need to be modified.

Comment (Thenerdstation) [reply tensornetwork/linalg/linalg.py]: My opinion is to leave it as `None` and just stick to the default of the given backend, as that default option will usually be the most preformant choice for that backend.

I doubt there would be many "obsure bugs" related to a bad dtype. Infact, its actually more likely that the obsure bug would be a slow down do to defaulting to float64 instead of float32 for pytorch since float32 runs significantly faster on GPU hardware.

My preference is to leave it as `None` instead of adding a default dtype to the backends that the user can't control.

Comment (alewis) [reply tensornetwork/linalg/linalg.py]: In this case, pending Martin's agreement, all issues with this PR appear to be resolved

Comment (mganahl) [reply tensornetwork/linalg/linalg.py]: I consider myself to be overuled :)
